src/pydocstyle.py

```diff
@@ -1696,18 +1700,93 @@ def check_starts_with_this(self, function, docstring):
 
-    # Somewhat hard to determine if return value is mentioned.
-    # @check(Function)
-    def SKIP_check_return_type(self, function, docstring):
-        """D40x: Return value type should be mentioned.
+    @check_for(Function)
+    def check_return_type(self, function, docstring):
+        """D405: Return value type should be mentioned.
 
         [T]he nature of the return value cannot be determined by
         introspection, so it should be mentioned.
 
         """
-        if docstring and function.returns_value:
-            if 'return' not in docstring.lower():
-                return Error()
+
+        class ReturnStatementTransformer(ast.NodeTransformer):
+
+            def __init__(self, *a, **kw):
+                self.value = None
+                self.parsed_one_function = False
+                super(ReturnStatementTransformer, self).__init__(*a, **kw)
+
+            def visit_Return(self, node):
+                # ignore empty "return"
+                if node.value is None:
+                    return
+                if isinstance(node.value, ast.Name):
+                    # ignore "return None"
+                    if node.value.id == 'None':
+                        return
+                    # document named return variables
+                    # return x
+                    #   D405: Return value type should be mentioned. ('x')
+                    self.value = node.value.id
+
+                # document all other return types
+                # return 5
+                #   D405: Return value type should be mentioned. ('Num')
+                # return (5, 6)
+                #   D405: Return value type should be mentioned. ('Tuple')
+                # return [5, 6]
+                #   D405: Return value type should be mentioned. ('List')
+                # return "t"
+                #   D405: Return value type should be mentioned. ('Str')
+                # return {"t": 5}
+                #   D405: Return value type should be mentioned. ('Dict')
+                # FIXME: this is using ast class names, could be nicer and
+                # translate into python names
+                self.value = node.value.__class__.__name__
+
+            def visit_FunctionDef(self, node):
+                # only parse the first FunctionDef
+                if self.parsed_one_function:
+                    return node
+                self.parsed_one_function = True
+                super(ReturnStatementTransformer, self).generic_visit(node)
+                return node
+
+            def visit_ClassDef(self, node):
+                # ignore all ClassDefs
+                pass
+
+        # ignore functions without docstrings
+        if not docstring:
+            return
+
+        # ignore functions that already mention return
+        # FIXME: this is only checking if "return" is anywhere in
+        # the docstring, not if the return is properly documented.
+        if 'return' in docstring.lower():
```

Comment:
I'm not certain this is as effective. I think having return somewhere in the docstring could lead to false positives (i.e., someone having the word "return" in their string without specifying `Returns` or `:returns:`. What if we look for those at the start of a line instead of just checking for `return` in the string?

Here are the three types of return documentation I think we can largely expect:

Sphinx default:

```
:returns: ...
```

or 

```
:returns:
    ...
```

Napoleon (+Sphinx)

```
Returns:
    ...
```

Or Twisted/Pydoctor:

```
@return: ...
```

I think these are the three pre-dominant styles that we'd like to accommodate in the default syntax. (Others are looking at Numpy/Scipy styles.) We could instead split the docstring on new lines and check `.startswith((sphinx_style, napoleon_style, twisted_style))`?